File: Leetcode/C++/332. Reconstruct Itinerary/solution.cpp
// https://leetcode.com/problems/reconstruct-itinerary/

#include <vector>
#include <algorithm>
#include <map>
#include <string>

using namespace std;

class Solution
{
public:
    vector<string> path;
    vector<string> findItinerary(vector<vector<string>> &tickets)
    {
        unordered_map<string, vector<string>> airports;
        // To check visit by using pop_back on a vector, sort it in bigger lexical order.
        sort(tickets.rbegin(), tickets.rend());
        for (auto ticket : tickets)
        {
            airports[ticket[0]].push_back(ticket[1]);
        }
        dfs(airports, "JFK");
        // Airports are added to the path starting from the last airport, it needs to be reverse
        reverse(path.begin(), path.end());
        return path;
    }

    void dfs(unordered_map<string, vector<string>> &airports, string start)
    {
        while (!airports[start].empty())
        {
            string nextAirport = airports[start].back();
            airports[start].pop_back();
            // Visit smaller lexical order airport
            dfs(airports, nextAirport);
        }
        path.push_back(start);
    }
};

File: Leetcode/Typescript/332. Reconstruct Itinerary/solution.ts
// https://leetcode.com/problems/reconstruct-itinerary/description/

function findItinerary(tickets: string[][]): string[] {
  let airport = new Map<string, string[]>();

  for (const airline of tickets) {
    const from = airline[0],
      to = airline[1];

    if (!airport.has(from)) {
      airport.set(from, []);
    }
    airport.get(from).push(to);
  }

  for (const array of airport.values()) {
    array.sort().reverse();
  }

  const stack = ["JFK"];
  const path: string[] = [];
  while (stack.length > 0) {
    const cur = stack[stack.length - 1];

    if (!airport.has(cur) || airport.get(cur).length === 0) {
      path.push(cur);
      stack.pop();
      continue;
    }

    stack.push(airport.get(cur).pop());
  }
  return path.reverse();
}
